Clients: Possibility to register replica after successful upload; Fix #1599

hahahannes opened the pull request:
Clients: Possibility to register replica after successful upload; Fix #1599

hahahannes committed:
Clients: Possibility to register replica after successful upload; Fix #1599

TWAtGH commented on lib/rucio/client/uploadclient.py:
```diff
@@ -153,11 +156,23 @@ def upload(self, items, summary_file_path=None):
+                if rsemgr.exists(rse_settings, file_did):
+                    logger.debug([replica for replica in self.client.list_replicas(dids=[{'name': file['did_name'], 'scope': file['did_scope']}])])
+
+                    if not [replica for replica in self.client.list_replicas(dids=[{'name': file['did_name'], 'scope': file['did_scope']}])]:
```
Maybe it's possible to replace list_replicas with a less expensive call?
Also please remove the double call for the debug output.

TWAtGH commented on lib/rucio/client/uploadclient.py:
```diff
@@ -153,11 +156,23 @@ def upload(self, items, summary_file_path=None):
+                    else:
+                        logger.info('File already registered. Skipping upload.')
+                        continue
+            elif not is_deterministic and not no_register and not register_after_upload:
```
`not register_after_upload` can be removed since the `elif` implies it

TWAtGH commented on lib/rucio/client/uploadclient.py:
```diff
@@ -131,6 +133,7 @@ def upload(self, items, summary_file_path=None):
             logger.info('Preparing upload for file %s' % basename)
 
             no_register = file.get('no_register')
+            register_after_upload = file.get('register_after_upload')
```
I think its better to force `register_after_upload = False` when `no_register = True`, e.g., `register_after_upload = file.get('register_after_upload') and not no_register`.

TWAtGH commented on lib/rucio/client/uploadclient.py:
```diff
@@ -153,11 +156,23 @@ def upload(self, items, summary_file_path=None):
-            if not is_deterministic and not no_register:
+            # if register_after_upload, file should be overwritten if it is not registered
+            # otherwise if file already exists on RSE we're done
+            if register_after_upload:
```
This should be fine if you force `register_after_upload = False` as suggested above

hahahannes committed:
Clients: added requested performance improvements and small changes; Fix #1599

bari12 commented on lib/rucio/client/uploadclient.py:
```diff
@@ -153,11 +156,23 @@ def upload(self, items, summary_file_path=None):
+                    try:
+                        dids = self.client.get_did(file['did_scope'], file['did_name'])
+                        logger.info('File already registered. Skipping upload.')
+                        logger.debug(dids)
```
Don't think we should leave the debug in?

bari12 commented on lib/rucio/client/uploadclient.py:
```diff
@@ -153,11 +156,23 @@ def upload(self, items, summary_file_path=None):
+                        continue
+                    except DataIdentifierNotFound:
+                        logger.info('File already exists on RSE. Existing replica will be removed first.')
+                        rsemgr.delete(rse_settings, [{'name': file['did_name'], 'scope': file['did_scope']}])
```
I don't think this workflow is correct.
rsemgr.upload(...) already takes care of the proper deletion if an incomplete file got uploaded (Check rsemanager.upload and the protocols). So you don't have to do this here. I think it might be enough to just call the upload and catch the exception and then do the registration

hahahannes replied:
Does it only count for incomplete files?
Because in my current test I upload a file and only remove it from the dids to force the overwrite, but it does not work, because the file already exists according to the posix exist() check.

hahahannes committed:
Clients: removed logging & deletion of left overs; Fix #1599

bari12 commented:
Sorry for the late follow up on the review. For me it looks good. 👍 @TWAtGH can you please have a look again as well.
The tests seem to fail right now though

```
======================================================================
FAIL: CLIENT(USER): Rucio upload files with registration after upload
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/opt/rucio/lib/rucio/tests/test_bin_rucio.py", line 270, in test_upload_file_register_after_upload
    nose.tools.assert_true((self.upload_success_str % path.basename(tmp_file4)) in out)
AssertionError: False is not true
```

hahahannes commented:
Yes, the test fails because of my simulation of a broken upload. I uploaded a file and removed the DID. When I then try to reupload the file it fails at the moment because the protocol existence check says that the file already exists. 
Before my last commit, I deleted the replica before reuploading it like this:
```python
logger.info('File already exists on RSE. Existing replica will be removed first.')
rsemgr.delete(rse_settings, [{'name': file['did_name'], 'scope': file['did_scope']}])
``` 
Should I put it back oder remove the test case? I was not sure if the test case was correct because it is not really a broken upload.

Edit: It's this [test](https://github.com/rucio/rucio/pull/1727/files#diff-90834bcf0a2db0761cdad97c256665dcR255)

bari12 commented:
I think the problem here is that posix is a renaming protocol. With a renaming protocol the file is first uploaded as `filename.upload` and once successful renamed to `filename`. How the rsemanager works right now, if something is already in it's final state (`filename`) it is considered to be done. Only `filename.upload` would be re-tried again.

On the other hand you are testing a valid case where you upload (+rename) and then the process crashes (and the post-registration is never done); Maybe it is valid to say in this case (`exception.FileReplicaAlreadyExists`) that we delete the replica and re-upload it, since it is not registered yet, but I am not 100% sure. @cserf any input?

cserf commented:
OK to delete and reupload. Maybe we can have another exception for this e.g. `FileBrokenUpload` or whatever better name.

bari12 commented:
Ok so in the `--registration-after-upload` case we delete the file, even if the full file (`filename`)already exists. I think however that this should be handled in the rsemanager than. Maybe with a new parameter in `upload` `delete_existing=True` or so. Not a huge fan of doing the deletion protocol handling directly in the uploadclient.

hahahannes committed:
Clients: delete existing files in case of register_after_upload and missing DID; Fix #1599

hahahannes committed:
Clients: also remove exisiting files that are already renamed; Fix #1599
